mico-core/src/main/java/io/github/ust/mico/core/REST/ApplicationController.java

```diff
@@ -155,6 +156,31 @@
         return ResponseEntity.ok(new Resource<>(updatedApplication, linkTo(methodOn(ApplicationController.class).updateApplication(shortName, version, application)).withSelfRel()));
     }
 
+    @DeleteMapping("/{" + PATH_VARIABLE_SHORT_NAME + "}/{" + PATH_VARIABLE_VERSION + "}")
+    public ResponseEntity<Resource<MicoApplication>> deleteApplication(@PathVariable(PATH_VARIABLE_SHORT_NAME) String shortName,
+                                                                       @PathVariable(PATH_VARIABLE_VERSION) String version) {
+        Optional<MicoApplication> applicationOptional = applicationRepository.findByShortNameAndVersion(shortName, version);
+        if (!applicationOptional.isPresent()) {
+            // application already deleted
+            return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+        }
+        applicationOptional.map(application -> {
+            if (application.getDeploymentInfo() == null || application.getDeploymentInfo().getServiceDeploymentInfos() != null) {
+                return application; // TODO better deployment detection
+            }
+            throw new ResponseStatusException(HttpStatus.CONFLICT, "Application is currently deployed!");
+        }).map(application -> {
+            return application.toBuilder().clearServices().build();
+        }).map(application -> {
+            applicationRepository.save(application);
+            return application;
+        }).map(application -> {
+            applicationRepository.delete(application);
+            return application;
+        });
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
```

Comment:
Why do we return `HttpStatus.NO_CONTENT`? Wouldn't it be better so return some `SUCCESS`?